### PR TITLE
Handle max binlog size exceeded error with smaller batch

### PR DESCRIPTION
**Pull request by Saadkhalid913**
We add a rescue clause into the chunker to handle only the error raised when binlog size is exceeded
We add a method to reduce stride size in throtter by a given % (default 20, configurable by user)
Backoff will keep happening until it reaches "MIN_STRIDE_SIZE" (set to 1 by now, not configurable at the moment)

### CODE REVIEW

File: spec/integration/chunker_spec.rb [hunk 1]
```diff
@@ -5,6 +5,7 @@
 require 'lhm/table'
 require 'lhm/migration'
 
+
```
[the remark this answers is not in the document]
Comment (Saadkhalid913) [reply spec/integration/chunker_spec.rb:8]: Removed!

**Review by coding-chimp — APPROVED** [review]

**Comment by Saadkhalid913** [comment]
Agent successfully ran migration with this version of LHM, and retried with smaller batch size and no errors, logs below
https://schema-migrations-staging.shopifysvc.com/schema-migrations-test-app/staging/main/20240815054742

File: lib/lhm/throttler/time.rb [hunk 2]
```diff
@@ -5,13 +5,46 @@ class Time
...
+          raise ArgumentError, 'min_stride_size must be less than or equal to stride'
+        end
+      end
+
+      def backoff_stride
+        new_stride = (@stride * (1 - @backoff_reduction_factor)).to_i
+
+        if new_stride == @stride
+          raise RuntimeError, "Cannot backoff any further"
+        end
```
Comment (chenhuiyeh) [inline lib/lhm/throttler/time.rb:42]: When would this happen? It makes sense looking at the logic, though I'm trying to think of potential cases:

* when `@backoff_reduction_factor` == 0 - This should be caught on initialize.
* when `@stride` == 0 - This should be caught on initialize (`@min_stride_size  >= 1 && @min_stride_size <= @stride`).
* Would rounding be the case here? An additional safeguard?

Curious to know, I may be missing something..

Comment (Saadkhalid913) [reply lib/lhm/throttler/time.rb:42]: You're correct about the rounding! With normal use there shouldn't actually be any situation in which this clause is invoked, it's just meant to be a safeguard against future changes that could potentially introduce an infinite loop. However it is possible to set stride, and backoff_reduction_factor in such a way that results in stride == new_stride (although you really have to be trying to break things)

Screenshot 2024-08-15 at 6 43 50 PM

File: lib/lhm/throttler/time.rb [hunk 3]
```diff
@@ -5,13 +5,46 @@ class Time
...
         @stride = options[:stride] || DEFAULT_STRIDE
+        @backoff_reduction_factor = options[:backoff_reduction_factor] || DEFAULT_BACKOFF_REDUCTION_FACTOR
+        @min_stride_size = options[:min_stride_size] || MIN_STRIDE_SIZE
+
+        if @backoff_reduction_factor >= 1 || @backoff_reduction_factor <= 0
+          raise ArgumentError, 'backoff_reduction_factor must be between greater than 0, and less than 1'
+        end
+
+        if @min_stride_size < 1
+          raise ArgumentError, 'min_stride_size must be and integer greater than 0'
```
[the remark this answers is not in the document]
Comment (Saadkhalid913) [reply lib/lhm/throttler/time.rb]: Fixed!

File: lib/lhm/throttler/time.rb (hunk 2 above)
Comment (chenhuiyeh) [reply lib/lhm/throttler/time.rb:42]: Thanks!

**Pull request merged**